src/OrleansRuntime/Streams/QueueBalancer/QueueBalancerBaseClass.cs

```diff
@@ -0,0 +1,72 @@
+﻿using Orleans.Providers;
+using Orleans.Runtime;
+using Orleans.Streams;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.Streams
+{
+    /// <summary>
+    /// Base class for StreamQueueBalancer
+    /// </summary>
+    public abstract class QueueBalancerBaseClass : IStreamQueueBalancer
+    {
+        protected readonly List<IStreamQueueBalanceListener> queueBalanceListeners;
+        public QueueBalancerBaseClass()
+        {
+            this.queueBalanceListeners = new List<IStreamQueueBalanceListener>();
+        }
+        public abstract IEnumerable<QueueId> GetMyQueues();
+        public abstract Task Initialize(string strProviderName, IStreamQueueMapper queueMapper, TimeSpan siloMaturityPeriod, IProviderConfiguration providerConfig);
+        public virtual bool SubscribeToQueueDistributionChangeEvents(IStreamQueueBalanceListener observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+            lock (queueBalanceListeners)
+            {
+                if (queueBalanceListeners.Contains(observer))
+                {
+                    return false;
+                }
+                queueBalanceListeners.Add(observer);
+                return true;
+            }
+        }
+        public virtual bool UnSubscribeFromQueueDistributionChangeEvents(IStreamQueueBalanceListener observer)
+        {
+            if (observer == null)
+            {
+                throw new ArgumentNullException("observer");
+            }
+            lock (queueBalanceListeners)
+            {
+                return queueBalanceListeners.Contains(observer) && queueBalanceListeners.Remove(observer);
+            }
+        }
+
+
+        public static List<string> GetActiveSilos(ISiloStatusOracle siloStatusOracle, ConcurrentDictionary<SiloAddress, bool> immatureSilos)
```

Comment:
Some of these concerns are implementation specific (like the concurrency issue). Could this be moved into the class which uses it instead, or is this something which every implementation of this class will need?